### ISSUE TITLE
BuildStream fails to detect missing source alias if already cached

### ISSUE THREAD

**Issue by Cynical-Optimist** [buildstream-migration/bst-staging#1119]
[See original issue on GitLab](https://gitlab.com/BuildStream/buildstream/-/issues/1119)
In GitLab by [[Gitlab user @palvarez89]](https://gitlab.com/palvarez89) on Sep 4, 2019, 16:19

## Summary

[//]: # (Summarize the bug encountered concisely)
If you use in your elements an alias, that isn't defined in your `project.conf`, and you happen to have that source cached, BuildStream will just use the cached source. This behaviour is confusing, as I had some builds working locally and only detected this error when sending my work to the CI.

## Steps to reproduce

[//]: # (How one can reproduce the issue - this is very important)
```bash
# 1 Clone freedesktop sdk
git clone https://gitlab.com/freedesktop-sdk/freedesktop-sdk.git
cd freedesktop-sdk

# 2 Force build glib element, to cache glib sources
vim element/components/glib.bst # Add 'echo 1' into the integration commands
bst build components/glib.bst

# 3 Remove 'gnome:' alias
sed -i -e 's/gnome.*//' include/aliases.yml

# 5 Force rebuild of glib, to reproduce the bug
vim element/components/glib.bst # Add 'echo 2' into the integration commands
bst build components/glib.bst

# 6 Remove cached glib source
find ~/.cache/buildstream/sources/ -name gnome_glib -exec rm -rf "{}" \;

# 7 Attempt to force rebuild glib
vim element/components/glib.bst # Add 'echo 3' into the integration commands
bst build components/glib.bst
```

## What is the current bug behavior?

[//]: # (What actually happens)
After step 5, BuildStream builds `glib` successfully.

## What is the expected correct behavior?

[//]: # (What you should see instead)
Step 5 should fail, as it doesn't have the `gnome:` alias defined.

In some cases, you might find yourself copying elements from one project to
another, after building some of these elements from the former. The new
project will not have the aliases needed for those elements, but the
build will succeed. You will only notice these failures when building
in a clean environment (e.g. CI)

## Relevant logs and/or screenshots

[//]: # (Paste any relevant logs - please use code blocks ``` to format console output, logs, and code as it's hard to read otherwise. You can also add a Snippet and link it here. Check the markdown giude on Gitlab for further tips)

## Possible fixes

[//]: # (If you can, link to the line of code that might be responsible for the problem)

## Other relevant information

[//]: # (Add the BuildStream version. Substitute below the "x" by "1", "2" or the right version. Add more than one version if necessary)

* BuildStream version affected: /milestone %"BuildStream_v1.4"

----
[//]: # (To review information about possible relevant labels for this issue please view the list of labels: https://gitlab.com/BuildStream/buildstream/labels)

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @cs-shadow]](https://gitlab.com/cs-shadow) on Oct 10, 2019, 19:29

changed the description

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @palvarez89]](https://gitlab.com/palvarez89) on Nov 15, 2019, 12:31

mentioned in commit celduin/crash/jetbot-system-bst[[Gitlab user @8dcce10293881b6e22073bbe8f5535dadaa41167]](https://gitlab.com/8dcce10293881b6e22073bbe8f5535dadaa41167)

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @cs-shadow]](https://gitlab.com/cs-shadow) on Dec 31, 2019, 15:09

Interesting find, thanks for reporting.

So, I think the problem comes from the fact that BuildStream is (at least currently) unable to distinguish between an untargetted alias and a legitimate URL. In the above example, until step 3 `gnome:` is expected to be an alias. After the alias is removed, BuildStream just thinks that the url is literally `gnome:path/to/repository`.

Treating it literally as an URL will actually even work in cases, where a host with that name is defined either in the ssh or git configuration. This may not behave in the same way for other sources where `:` isn't usually expected.

Since the full expanded url is not part of the cache key, BuildStream produces the same cache key with and without the alias. And since the sources are fetched in step 2 already, everything works fine in the local case.

---

For the fix, I can think of a few options:

1. Include the full URL in the cache key - this will break cache keys in cases where user wants to change the target url of an alias.

2. Include whether or not an alias was used when calculating the cache key. This should allow us to differentiate between aliased URLs and URLs having `:` in them - this will still break cache key once but not while changing target urls.

3. Reserve the `:` character for aliases and disallow them from various source fields. Perhaps by providing some mechanism to escape it. - This won't break cache keys but will introduce a new restriction on source metadata.

Not sure which is the most correct solution just yet. Suggestions welcome.
